File: day24/python/bridge.py
#!/usr/bin/python3

import sys

bridges = []

def calcStrength(component):
    return component[0] + component[1]

def buildBridge(parts, port, strength, length):
    bridges.append((strength, length))
    connectors = [c for c in parts if c[0] == port or c[1] == port]
    for c in connectors:
        tmp = list(parts)
        tmp.pop(tmp.index(c))
        if c[0] == port:
            buildBridge(tmp, c[1], strength + calcStrength(c), length + 1)
        else:
            buildBridge(tmp, c[0], strength + calcStrength(c), length + 1)

if len(sys.argv) != 2:
    print('Provide input file as command line argument!')
else:
    components = open(sys.argv[1]).read().splitlines()
    components = [(int(c.split('/')[0]), int(c.split('/')[1])) for c in components]

    buildBridge(components, 0, 0, 0)
    print('Strongest:', max(bridges, key = lambda sl: sl[0])) # print the strongest

    longest = []
    bridge = bridges.pop(bridges.index(max(bridges, key = lambda sl: sl[1])))
    longest.append(bridge)
    while max(bridges, key = lambda sl: sl[1])[1] == bridge[1]:
        bridge = max(bridges, key = lambda sl: sl[1])
        longest.append(bridges.pop(bridges.index(bridge)))

    print('Strongest of the longest:', max(longest, key = lambda sl: sl[0])) # print the longest

File: day24/rust/bridge.rs
use std::fs;

struct Bridge {
    strength: usize,
    length: usize,
}

fn build_bridge(
    bridges: &mut Vec<Bridge>,
    components: &Vec<(usize, usize)>,
    port: usize,
    strength: usize,
    length: usize,
) {
    bridges.push(Bridge { strength, length });
    let connectors: Vec<usize> = components
        .iter()
        .filter(|component| component.0 == port || component.1 == port)
        .map(|component| {
            if component.0 == port {
                component.1
            } else {
                component.0
            }
        }).collect();

    for connector in connectors {
        let index = components
            .iter()
            .position(|component| {
                (component.0 == connector && component.1 == port)
                    || (component.0 == port && component.1 == connector)
            }).unwrap();
        let mut reduced_components = components.to_vec();
        reduced_components.remove(index);

        let calculate_strength = |component: (usize, usize)| component.0 + component.1 + strength;
        build_bridge(
            bridges,
            &reduced_components,
            connector,
            calculate_strength(components[index]),
            length + 1,
        );
    }
}

fn main() {
    let content = fs::read_to_string("../input").expect("file not found");
    let mut components = Vec::new();
    let mut bridges = Vec::new();

    for line in content.lines() {
        let mut ports = line.split('/');
        components.push((
            ports.next().unwrap().parse::<usize>().unwrap(),
            ports.next().unwrap().parse::<usize>().unwrap(),
        ));
    }

    build_bridge(&mut bridges, &components, 0, 0, 0);

    let max_strength = bridges
        .iter()
        .max_by(|&bridge_1, &bridge_2| (*bridge_1).strength.cmp(&bridge_2.strength))
        .unwrap()
        .strength;
    println!("{}", max_strength);

    let strength_of_the_longest = bridges
        .iter()
        .max_by(|&bridge_1, &bridge_2| (*bridge_1).length.cmp(&bridge_2.length))
        .unwrap()
        .strength;
    println!("{}", strength_of_the_longest);
}
